### DISCUSSION TITLE
Documentation on DebugMode

### DISCUSSION THREAD

**Discussion by OSG-Pierre** [sebastienros/jint#1236] [Q&A]
Hi,
I'm trying to use the debug mode but I can't seem to find any documentation on how to operate it. For example I got this little code running to log each step of my script:

        jintEngine.Step += (sender, info) =>
        {
            Debug.Log($"{info.CurrentStatement}: Line {info.CurrentStatement.Location.Start.Line}, Col {info.CurrentStatement.Location.Start.Column}, identifier: {identifier}");
            return Jint.Runtime.Debugger.StepMode.Into;
        };

But then I wanted to know the name of the function I was currently debugging. This is where a documentation would help me a lot. So far I'm doing it like follows, but I have no idea how to get variable names, etc.

        jintEngine.Step += (sender, info) =>
        {
            string identifier = "";
            var funcStatement = info.CurrentStatement.As<Jint.Parser.Ast.FunctionDeclaration>();

            if (funcStatement != null)
            {
                identifier = funcStatement.Id.Name;
            }

            Debug.Log($"{info.CurrentStatement}: Line {info.CurrentStatement.Location.Start.Line}, Col {info.CurrentStatement.Location.Start.Column}, identifier: {identifier}");
            return Jint.Runtime.Debugger.StepMode.Into;
        };

So is there a documentation on how to use the debug mode? Or examples?
Thank you for your help :)

**Comment by Jither** [comment]
I think the main issue here, before it makes sense to document it, is that `DebugMode` isn't fully functional. Main issues I've identified - and am trying to fix, likely for a future pull request, when it has been tested further:

- block scoped (const/let) variables aren't reported in `Globals` or `Locals` (fixed, I think)
- `Globals` and `Locals` aren't separated logically - i.e., Globals includes all local variables (fixed, I think)
- missing support for `Source` in breakpoints (fixed, I think)
- any callee that isn't referenced with a simple `Identifier` will not be added to the call stack (fixed, I think - it now does the same function naming as the Chromium debugger - using Jint's own function name inference). The argument list has been removed from the call stack, since: it's not all that useful; it requires evaluating the arguments, or making special cases to stringify every expression that might be passed, or using Esprima to stringify it; and no other Javascript debuggers I know of include the argument list, probably for much the same reasons). The values of arguments can always be found in `Locals` anyway.
- some constructs would cause the debugger to crash (fixed the ones I've seen so far)
- `StepMode` behaves very erratically (looking into that now)

The baseline right now is, "What Would Chromium or Firefox do?"

With the fixes thus far, without any changes to the Jint `DebugMode` API itself (other than adding Source property to Breakpoints), and with a minimal set of changes to Jint other than in the `Jint.Runtime.Debugger` namespace<sup>1</sup>, I now have a WPF-based debugging client that does most of what you'd expect from a JS debugger, but unfortunately it can't be shared, since it has company-owned source code dependencies (even if I wrote those too). Still, I might be able to rip it out of the grasp of those dependencies as an example, or maybe turn it console based for more cross platform and less boilerplate, if the pull request comes through.

<sup>1</sup> Actually, so far, outside the Debugger namespace there's just one moved conditional call, and one function rewritten - a function that is only used by the current `DebugHandler` code in the first place.
